Reworked error output to send IETF RFC-7807 compatible problem responses

Alexamakans opened the pull request:
Closes  #15
Depends on https://github.com/iver-wharf/wharf-core/pull/17

Alexamakans committed:
Use ginutil.WriteProblemError instead of directly using gin.Context methods

Alexamakans committed:
Updated CHANGELOG.md

Alexamakans committed:
Fixed Codacy analysis warning

Alexamakans committed:
Merge branch 'master' into feature/use-wharf-core-Problem

applejag commented on CHANGELOG.md:
```diff
@@ -39,6 +39,9 @@ This project tries to follow [SemVer 2.0.0](https://semver.org/).
 
 - Fixed importing from a group different than the authenticated user. (#13)
 
+- Changed to use ginutil.WriteProblemError instead of directly using
+  gin.Context methods. (#16)
```
nit: "But what does this mean for the end user?"

Your changes here actually changes the response structures when there's an error. Suggest specifying that, such as:

```suggestion
- Changed to return IETF RFC-7807 compatible problem responses on failures
  instead of solely JSON-formatted strings. (#16)
```

applejag commented on github.go:
```diff
@@ -36,8 +38,8 @@ func runGitHubHandler(c *gin.Context) {
 	if err != nil {
-		c.Error(err)
-		c.JSON(http.StatusBadRequest, err.Error())
+		ginutil.WriteInvalidBindError(c, err,
+			"One or more parameters failed to parse when reading the request body for GitHub projects import/refresh")
```
Remember to update the `// @failure` response type above

```diff
-// @Failure 400 {object} string "Bad request"
-// @Failure 401 {object} string "Unauthorized or missing jwt token"
+// @Failure 400 {object} problem.Response "Bad request"
+// @Failure 401 {object} problem.Response "Unauthorized or missing jwt token"
```

applejag commented on github.go:
```diff
@@ -54,37 +56,57 @@ func runGitHubHandler(c *gin.Context) {
+		ginutil.WriteProblemError(c, err, problem.Response{
+			Type: "prob/provider/github/getting-token-error",
+			Title: "Error getting token.",
+			Status: http.StatusBadRequest,
```
Would declare this more as a "Bad Gateway" error, as the error occurred when acting as a gateway to the API via the `importer.getToken` method.

applejag replied:
Same comment for the other problems

applejag commented on github.go:
```diff
@@ -54,37 +56,57 @@ func runGitHubHandler(c *gin.Context) {
+			Type: "prob/provider/github/getting-token-error",
+			Title: "Error getting token.",
+			Status: http.StatusBadRequest,
+			Detail: "Unable to get token.",
```
Suggest adding some details here about what token, such as

```suggestion
			Detail: fmt.Sprintf("Unable to get token by ID %d or create token new token.", i.TokenID),
```

applejag replied:
Same comment for the other problems. The problem detail should include lots of THIS-PROBLEM-INSTANCE-related data, to make it easier to understand what's going on.

Like, the problem detail should be able to explain the error itself, as the error field is not actually part of the IETF RFC-7807 spec.

So almost include as much info as possible, as long as you're not including any confidential or sensitive data (such as the token value)

Alexamakans replied:
Understood.

I will try to include more info for the errors tomorrow. I need to go through the code a bit more to understand the specifics.

applejag replied:
I wrote a lot over at https://github.com/iver-wharf/wharf-provider-azuredevops/pull/14

Instead of repeating myself, try to use that one as a template to how to solve stuff for this one

applejag commented on github.go:
```diff
@@ -54,37 +56,57 @@ func runGitHubHandler(c *gin.Context) {
-		c.Error(err)
-		c.JSON(http.StatusBadRequest, fmt.Sprintf("Unable to get provider. %+v", err))
+		ginutil.WriteProblemError(c, err, problem.Response{
+			Type: "prob/provider/github/getting-provider-error",
```
Missing leading slash in Type

```suggestion
			Type: "/prob/provider/github/getting-provider-error",
```

applejag commented on github.go:
```diff
@@ -54,37 +56,57 @@ func runGitHubHandler(c *gin.Context) {
-		c.Error(err)
-		c.JSON(http.StatusBadRequest, fmt.Sprintf("Unable to init github connection. %+v", err))
+		ginutil.WriteProblemError(c, err, problem.Response{
+			Type: "prob/provider/github/connection-error",
```
Missing leading slash in Type

```suggestion
			Type: "/prob/provider/github/connection-error",
```

applejag commented on github.go:
```diff
@@ -54,37 +56,57 @@ func runGitHubHandler(c *gin.Context) {
-			c.Error(err)
-			c.JSON(http.StatusBadRequest, fmt.Sprintf("Unable to import project. %+v", err))
+			ginutil.WriteProblemError(c, err, problem.Response{
+				Type: "prob/provider/github/import-project-error",
```
Missing leading slash in Type

```suggestion
				Type: "/prob/provider/github/import-project-error",
```

applejag commented on github.go:
```diff
@@ -54,37 +56,57 @@ func runGitHubHandler(c *gin.Context) {
-			c.Error(err)
-			c.JSON(http.StatusBadRequest, fmt.Sprintf("Unable to import organization or group. %+v", err))
+			ginutil.WriteProblemError(c, err, problem.Response{
+				Type: "prob/provider/github/import-group-error",
```
Missing leading slash in Type

```suggestion
				Type: "/prob/provider/github/import-group-error",
```

applejag commented on github.go:
```diff
@@ -2,6 +2,8 @@ package main
 
 import (
 	"fmt"
+	"github.com/iver-wharf/wharf-core/pkg/ginutil"
+	"github.com/iver-wharf/wharf-core/pkg/problem"
```
nit: Could you move these to down near the other `github.com/iver-wharf/...` imports?

Alexamakans committed:
Update CHANGELOG.md

Co-authored-by: kalle (jag) <kalle.jillheden@iver.se>

Alexamakans committed:
Added suggestions

- Added some more information in Detail for some errors.
- Changed CHANGELOG.md entry.
- Changed Failure type.
- Changed some errors to be Bad Gateway instead.
- Fixed missing leading slashes in Type.
- Organized imports.

Alexamakans committed:
Merge branch 'feature/use-wharf-core-Problem' of github.com:iver-wharf/wharf-provider-github into feature/use-wharf-core-Problem

Alexamakans committed:
Fixed errors

applejag reviewed (changes requested):
Wrote a lot of suggestions over at https://github.com/iver-wharf/wharf-provider-azuredevops/pull/14

Instead of repeating myself I suggest looking at those comments to use as template for this one

Alexamakans committed:
Changed to use generic problem types with better details

Alexamakans committed:
Sort imports and fixed a syntax error

Alexamakans commented on helpers/ginutilext/extension.go:
```diff
@@ -0,0 +1,63 @@
+package ginutilext
```
I also have the this file in my local branch for wharf-provider-azuredevops right now.

My reasoning is that it makes it easier to move to wharf-core/pkg/ginutil, once they're more decided.

I couldn't find a good way to import a branch of a repository in Go.
If there is a way to do that then I could just make a branch for wharf-core and import that until the problems are ready to be pushed.

Not sure if this is a good way to do it, would love to hear some other solutions.

fredx30 replied:
It seems like this might be solved by the merge order. If you cross link all the relevant PRs and number them by merge order. Then resolve this comment at the end just before this is ready for merge. If this isnt ready for merge yet because dependencies are missing you could switch this to a draft for now.

Alexamakans committed:
fixup! Changed to use generic problem types with better details

Alexamakans committed:
Added comments to ginutilext/extension.go

Alexamakans committed:
Changed to use wharf-core/pkg/ginutil instead of pkg/ginutilext

Alexamakans committed:
Merge remote-tracking branch 'origin' into feature/use-wharf-core-Problem

Alexamakans committed:
Updated CHANGELOG.md

Alexamakans committed:
Updated CHANGELOG.md

applejag commented on github.go:
```diff
@@ -10,6 +10,8 @@ import (
 	"github.com/google/go-github/github"
 	"github.com/iver-wharf/wharf-api-client-go/pkg/wharfapi"
+	"github.com/iver-wharf/wharf-core/pkg/ginutil"
+	_ "github.com/iver-wharf/wharf-core/pkg/problem"
```
Why was this import for `github.com/iver-wharf/wharf-core/pkg/problem` added as a discard import?

Alexamakans replied:
It seemed like it was necessary to have it imported so that `problem.Response` would be recognized as a type in e.g.
```
// @Failure 400 {object} problem.Response "Bad request"
// @Failure 401 {object} problem.Response "Unauthorized or missing jwt token"
// @Failure 502 {object} problem.Response "Bad gateway"
```
but since that wasn't recognized as usage in my GoLand I had to make it a discard import for it to be happy.

Alexamakans replied:
Now that I'm using VS Code it doesn't seem to be an issue, allowing me to leave it out completely.

applejag replied:
If you remove it, can you still do `swag init`? Maybe it was an issue with when the wharf-core module was not referenced at all but since you're referencing the `ginutil` that in turn references the `problem` package maybe that's enough?

Alexamakans replied:
Right!
Yeah it wouldn't work when I did swag init before, same thing now. It complains about not finding the type definition for problem.Response.

It does work with `--parseDepth=2` though, in which case I'm assuming it finds it through the `ginutil` package.

applejag replied:
Then I suggest you update the README and rely on it finding it via the `ginutil` package.

applejag commented on github.go:
```diff
@@ -29,15 +31,16 @@ type githubImporter struct {
 	if err != nil {
-		c.Error(err)
-		c.JSON(http.StatusBadRequest, err.Error())
+		ginutil.WriteInvalidBindError(c, err,
```
Need to use `c.ShouldBindJSON` now that you're using problem responses

https://iver-wharf.github.io/#/development/gin-shouldbind-vs-bind

applejag commented on github.go:
```diff
@@ -52,59 +55,69 @@ func runGitHubHandler(c *gin.Context) {
 }
 
-func (importer githubImporter) getToken(i importBody) (wharfapi.Token, error) {
+func (importer githubImporter) getToken(c *gin.Context, i importBody) (wharfapi.Token, error) {
```
Suggest renaming this method to state that it's writing problems such as `getTokenWritesProblem`, and suggest returning a boolean instead of an error as the errors are handled already.

Alexamakans committed:
Applied suggestions from code review

applejag commented on github.go:
```diff
@@ -52,63 +55,74 @@ func runGitHubHandler(c *gin.Context) {
-			c.Error(err)
-			c.JSON(http.StatusBadRequest, fmt.Sprintf("Unable to import project. %+v", err))
+			ginutil.WriteAPIClientWriteError(c, err,
+				fmt.Sprintf("Unable to import project %q (id: %v) from GitHub.", i.Project, i.ProjectID))
```
nit:

```suggestion
				fmt.Sprintf("Unable to import project %q with ID %d from GitHub.", i.Project, i.ProjectID))
```

applejag commented on github.go:
```diff
@@ -52,63 +55,74 @@ func runGitHubHandler(c *gin.Context) {
-			return token, err
+			ginutil.WriteAPIClientReadError(c, err,
+				fmt.Sprintf(
+					"Unable to get token by id %v. Likely because of a failed request or malformed response.",
```
nit:

```suggestion
					"Unable to get token by ID %d. Likely because of a failed request or malformed response.",
```

applejag commented on github.go:
```diff
@@ -52,63 +55,74 @@ func runGitHubHandler(c *gin.Context) {
+					i.TokenID))
 		} else if token.TokenID == 0 {
-			err = fmt.Errorf(fmt.Sprintf("Token with id %v not found", i.TokenID))
+			err = fmt.Errorf("token with id %v not found", i.TokenID)
```
nit:

```suggestion
			err = fmt.Errorf("token with ID %d not found", i.TokenID)
```

applejag commented on github.go:
```diff
@@ -52,63 +55,74 @@ func runGitHubHandler(c *gin.Context) {
-			err = fmt.Errorf(fmt.Sprintf("Token with id %v not found", i.TokenID))
+			err = fmt.Errorf("token with id %v not found", i.TokenID)
+			ginutil.WriteAPIClientReadError(c, err,
+				fmt.Sprintf("Token with id %v not found", i.TokenID))
```
nit:

```suggestion
				fmt.Sprintf("Token with ID %d not found.", i.TokenID))
```

applejag commented on github.go:
```diff
@@ -52,63 +55,74 @@ func runGitHubHandler(c *gin.Context) {
+		if err != nil {
+			ginutil.WriteAPIClientWriteError(c, err,
+				fmt.Sprintf(
+					"Unable to create token for %q. Likely because of a failed request or malformed response.",
```
nit: Good to always describe the values, as values such as an empty string will not be self explanatory.

```suggestion
					"Unable to create token for user %q. Likely because of a failed request or malformed response.",
```

applejag commented on github.go:
```diff
@@ -52,63 +55,74 @@ func runGitHubHandler(c *gin.Context) {
 
 	fmt.Println("Token from db: ", token)
-	return token, err
+	return token, err == nil
```
Suggest returning after each `ginutil.Write...` call instead, so it follows the same convention everywhere. Then you can assume this value to be nil here and just return the literal `nil` instead.

applejag reviewed (approved):
Some nits and a minor comment. Otherwise: LGTM!

Alexamakans committed:
Removed linebreak

Alexamakans committed:
Updated README.md

Alexamakans committed:
Applied suggestions from code review

Co-authored-by: kalle (jag) <kalle.jillheden@iver.se>

Alexamakans committed:
Applied suggestion from code review

Alexamakans committed:
Merged with origin/master
